design-document.md

```diff
@@ -11,9 +11,98 @@
-
-## Threat model
+## Architecture overview
+
+Prio is a system and protocol for privately computing aggregation functions over private 
+input. An aggregation function F is one that computes an output y = F(x[1],x[2],...) for inputs
+x[i]. In general, Prio supports any aggregation function whose inputs can be encoded in a 
+particular way. However, not all aggregation functions admit an efficient encoding, rendering
+them impractical to implement. Thus, Prio supports a limited set of aggregation functions, 
+some of which we highlight below:
+
+- Simple statistics, including sum, mean, min, max, variance, and standard deviation; 
+  [[OPEN ISSUE: It's possible to estimate quantiles such as the median. How practical is this?]]
+- Bit vector OR and AND operations; and
+- Data structures, like Bloom filters, counting Bloom filters, and count-min sketches, that 
+  approximately represent (multi-)sets of strings.
+
+The applications for such aggregations functions are large, including, though not limited to:
+counting the number of times a sensitive or private event occurs and approximating the frequency
+that sensitive tokens or strings occur.
+
+Client applications hold private inputs to the aggregation function, server processors,
+or aggregators, run a protocol that validates each input x[1], x[2], ... and computes the
+final output y. The final collector obtains the output of the aggregation function.
+
+At a high level, the flow of data through these entities works roughly as follows:
+
+~~~
+                            +------------+     
+                            |            |        
+                            | Aggregator |
+                            |            |
+                            +-^-------^--+
+                              |       |   
+                          (2) |       | (3)
+                              |       |    
++--------+           +--------v---+   |        +-----------+
+|        |    (1)    |            <---+   (4)  |           |
+| Client +----------->   Leader   +------------> Collector |
+|        |           |            <---+        |           |
++--------+           +--------^---+   |        +-----------+
+                              |       |       
+                          (2) |       | (3)       
+                              |       |        
+                            +-v-------v--+     
+                            |            |     
+                            | Aggregator |
+                            |            |
+                            +------------+
+~~~ 
+
+1. Upload: Clients split inputs into s >= 2 shares, encrypt each share for a different 
+   Aggregator, and send these encrypted shares to the Leader. Clients continue 
+   this process until a "batch" of data is collected.
+2. Verify: Upon receipt of an encrypted share, the Leader sends it to the respective 
+   Aggregator. Each Aggregator decrypts the share, computes a proof from the respective
```

Comment:
That's one possibility, sure, but I don't think that's necessary. The leader could package up signed keys that it collects from aggregators and send them to the client. Clients could also fetch them directly from aggregators (if they knew what aggregators to fetch from). From a client perspective, the former seems simpler, since it would allow us to punt on aggregator discovery.

Reply:
The leader could vend signed keys on behalf of aggregators, but the client still needs to discover a trusted public key for each aggregator to then verify the signed keys. I think you have to bootstrap trust between the client and each participating aggregator independently from the leader, or the leader can impersonate aggregators and defeat the system's privacy guarantees.

Reply:
IMO it's better to avoid creating a new PKI and just have aggregators distribute their own encryption keys. I'm open to being wrong, however :)

Reply:
I'm not suggesting a new PKI for this.

Reply:
>  but the client still needs to discover a trusted public key for each aggregator to then verify the signed keys.

Yes, it does, and I'm suggesting this can be done by asking the leader to provide the signed keys.